### PR TITLE
calculate and show centre of AirQloud

### PR DESCRIPTION
**Pull request by Baalmart**
# calculate and show centre of AirQloud

**_WHAT DOES THIS PR DO?_**

- [x] View the center of an AirQloud when viewing AirQloud results
- [x] Generate the center of AirQloud during creation process
- [x] Update AirQloud center during **AirQloud refresh** process

**_WHAT JIRA STORY/TASK IS RELATED TO THIS PR?_**
[PLAT-1016]

**_HOW DO I TEST OUT THIS PR?_**
README, development environment

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

- [x] [Find AirQloud center](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/airqlouds#find-center-of-airqloud)
- [x] [Create AirQloud](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/airqlouds#create-airqloud)
- [x] [Refresh AirQloud](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/airqlouds#refresh-airqloud)

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**

N/A

**_SCREENSHOTS?_**

Screenshot 2022-03-04 at 16 27 00
Screenshot 2022-03-04 at 16 29 37
Screenshot 2022-03-04 at 16 29 59
Screenshot 2022-03-04 at 16 31 25
Screenshot 2022-03-04 at 18 19 04

[PLAT-1016]: https://airqoteam.atlassian.net/browse/PLAT-1016?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/device-registry/controllers/create-airqloud.js [hunk 1]
```diff
@@ -63,6 +63,69 @@ const createAirqloud = {
     }
   },
 
+  calculateGeographicalCenter: async (req, res) => {
```
Comment (Baalmart) [inline src/device-registry/controllers/create-airqloud.js:66]: new controller logic for calculating the center of an AirQloud

File: src/device-registry/controllers/create-airqloud.js [hunk 2]
```diff
@@ -158,6 +221,10 @@ const createAirqloud = {
 
   findSites: async (req, res) => {
     try {
```
Comment (Baalmart) [inline src/device-registry/controllers/create-airqloud.js:223]: Temporarily disabling the endpoint for finding Sites within any given AirQloud. Done due to failure to resolve a runtime error.

File: src/device-registry/models/Airqloud.js [hunk 3]
```diff
@@ -50,9 +50,20 @@ const metadataSchema = new Schema(
   { _id: false }
 );
 
+const pointSchema = new Schema(
```
Comment (Baalmart) [inline src/device-registry/models/Airqloud.js:53]: Updating the schema to cater for the new field (center_point) which will handle the center of an AirQloud

File: src/device-registry/routes/api-v1.js [hunk 4]
```diff
@@ -3659,4 +3659,80 @@ router.delete(
   airqloudController.delete
 );
 
+router.post(
+  "/airqlouds/center",
```
Comment (Baalmart) [inline src/device-registry/routes/api-v1.js:3663]: New route for getting the center of an AirQloud. Most of the logic in here is just input validation.

File: src/device-registry/utils/create-airqloud.js [hunk 5]
```diff
@@ -32,16 +32,20 @@ const createAirqloud = {
       );
     }
   },
-  retrieveLocationCoordinates: async (request) => {
+
+  retrieveCoordinates: async (request, entity) => {
     try {
```
Comment (Baalmart) [inline src/device-registry/utils/create-airqloud.js:37]: refactored the util for retrieving coordinates of either a location or AirQloud

File: src/device-registry/utils/create-airqloud.js [hunk 6]
```diff
@@ -91,33 +95,50 @@ const createAirqloud = {
   },
   create: async (request) => {
     try {
-      const { body } = request;
-      const { tenant } = request.query;
-      const { location_id } = request.body;
+      const { body, query } = request;
+      const { tenant } = query;
```
Comment (Baalmart) [inline src/device-registry/utils/create-airqloud.js:99]: enhanced AirQloud creation with logic for adding the AirQloud center in the process

File: src/device-registry/utils/create-airqloud.js [hunk 7]
```diff
@@ -272,49 +291,88 @@ const createAirqloud = {
...
-          return {
-            success: false,
-            message: responseFromUpdateAirQloud.message,
-            status,
-            errors,
-          };
-        }
+
+      let requestForUpdateAirQloud = request;
+      requestForUpdateAirQloud["body"] = {};
```
Comment (Baalmart) [inline src/device-registry/utils/create-airqloud.js:298]: resolved runtime for AirQloud refresh. Just disabled the part of adding existing Sites and also add function to add center of an AirQloud. The part of adding Sites disabled due to unclear runtime errors

File: src/device-registry/utils/create-airqloud.js [hunk 8]
```diff
@@ -329,11 +387,63 @@ const createAirqloud = {
       };
     }
   },
+
+  calculateGeographicalCenter: async (request) => {
+    try {
+      const { body, query } = request;
+      const { coordinates } = body;
+      const { id } = query;
```
Comment (Baalmart) [inline src/device-registry/utils/create-airqloud.js:395]: Util logic for calculating the geographical center of an AirQloud

**Review by 123MwanjeMike — APPROVED** [review]

**Review by ja-odur — APPROVED** [review]
LGTM

**Pull request merged**